In [25]:
import sympy as sp

In [26]:
X = sp.Matrix([
  [1, 2],
  [5, 3],
  [2, 4],
  [3, 7]
])
X

Matrix([
[1, 2],
[5, 3],
[2, 4],
[3, 7]])

In [27]:
B = sp.Matrix([5, 6, 10, 8])
B

Matrix([
[ 5],
[ 6],
[10],
[ 8]])

In [28]:
psydo_inverse = (X.T * X) ** -1 * X.T
psydo_inverse

Matrix([
[-7/463,  126/463, -14/463, -44/463],
[16/463, -113/926,  32/463, 135/926]])

In [29]:
A = psydo_inverse * B
A

Matrix([
[229/463],
[601/463]])

In [30]:
mse = X * A - B
mse = mse.T * mse / mse.rows
float(mse[0])

6.241900647948164

In [31]:
A.applyfunc(float)

Matrix([
[0.494600431965443],
[ 1.29805615550756]])

Градиентный спуск в задании 9 дал $mse = 6.35$ в точке $(0.68, 1.17)$. Метот псевдообратных матриц улучшил результат $mse$ до $6.24$ в точке $(0.49, 1.30)$.
